### PR TITLE
[caffe2] Fix observer logic in RNN executor. Remove dynamic casts

### PR DESCRIPTION
**Pull request by mdschatz**
Test case failed that triggered PR#2427 as a bandaid. As part of this, we discussed redesigning how RNN-Capable observers were treated and landed on specifying a boolean for this purpose. Updated code to support this.

### CODE REVIEW

File: caffe2/operators/rnn/recurrent_network_executor.h [hunk 1]
```diff
@@ -161,10 +161,10 @@ class RecurrentNetworkExecutorBase {
...
-              auto rnn_observer =
-                  dynamic_cast_if_rtti<const RNNCapableOperatorObserver*>(
-                      observer.get());
-              if (rnn_observer) {
+              if (observer->IsRNNCapable()) {
+                auto rnn_observer =
+                    static_cast<const RNNCapableOperatorObserver*>(
+                        observer.get());
                 std::unique_ptr<ObserverBase<OperatorBase>> rnn_observer_copy =
                     rnn_observer->rnnCopy(rnn_op.op.get(), rnn_op.order);
```
Comment (salexspb) [inline caffe2/operators/rnn/recurrent_network_executor.h]: hmm, at this point an observer will have both set the flag, inherit from RNNCapableOperatorObserver and implement rnnCopy method. At this point we can just fall back to the approach with rnnCopy being in the base class, we leak notion of RNN to the base class in any way. Then we don't need any casting business. I know, this was your original approach :)

Comment (mdschatz) [reply caffe2/operators/rnn/recurrent_network_executor.h]: So then is the recommendation is to have a default implementation for rnnCopy that returns null if it's not capable?

Comment (mdschatz) [reply caffe2/operators/rnn/recurrent_network_executor.h]: And just so we don't re-re-design this, I recall one of the motivations for separating these two was to provide a cleaner abstraction between different levels of observers (Net, Operator, Step Net).  I want to make sure these concerns are resolved by reverting to the original design where we simply implemented the rnnCopy method (potentially returning NULL).

Comment (salexspb) [reply caffe2/operators/rnn/recurrent_network_executor.h]: This true, being rnn capable only applies to nets.
So this argument could applied to IsRNNCapable() method in the base class, not just copyForRNN method, IMO these approaches are not much different. 

I feel the main issue here is that all the observer framework was not meant for dealing with RNNs and now we have to pay for it when we would like to reuse observers inside an RNN :)

**Review by salexspb — APPROVED** [review]

**Review by houseroad — APPROVED** [review]
Thanks a lot for fixing this. :-)

**Pull request merged**